modules/consentManagementGpp.js

```diff
@@ -0,0 +1,430 @@
+        processCmpData(gppData, {onSuccess, onError});
+      } else if (evt.pingData.cmpStatus === 'error') {
+        onError('CMP returned with a cmpStatus:error response.  Please check CMP setup.');
+      } else if (!provisionalConsent) {
+        const provGppData = cmpFrame[cmpApiName]('getGPPData');
+        logInfo('Called GPP CMP early to determine the value of applicableSections; value returned from CMP: ', provGppData);
+        if (checkApplicableSectionsIsReady(provGppData)) {
+          provisionalConsent = provGppData;
+        } else if (checkApplicableSectionIsReady(provGppData)) {
+          provisionalConsent = provGppData;
+          // conform naming convention from old version of the spec to new version
+          provisionalConsent.applicableSections = provGppData.applicableSection;
+          delete provisionalConsent.applicableSection;
+        }
+      }
+    }
+  }
+
+  function callCmpWhileInIframe(commandName, cmpFrame, moduleCallback, cmpCallParam) {
+    // changing the name away from the standard, to properly handle multiple calls to CMP,
+    // otherwise follow-up calls made to fetch getGPPData routed here will complain later that there's no callback
+    const cmpApiPMName = '__gpp-pm';
+    const callName = `${cmpApiName}Call`;
+
+    /* Setup up a __cmp function to do the postMessage and stash the callback.
+    This function behaves (from the caller's perspective identically to the in-frame __gpp call */
+    window[cmpApiPMName] = function (cmd, callback, param) {
+      const callId = Math.random().toString();
+      const msg = {
+        [callName]: {
+          command: cmd,
+          parameter: param,
+          version: 1,
+          callId: callId
+        }
+      };
+
+      // TODO? - add logic to check if random was already used in the same session, and roll another if so?
+      cmpCallbacks[callId] = callback;
+      cmpFrame.postMessage(msg, '*');
+    }
+
+    // when we get the return message, call the stashed callback;
+    // register the listener only once per session
+    if (!registeredPostMessageResponseListener) {
+      window.addEventListener('message', readPostMessageResponse, false);
+      registeredPostMessageResponseListener = true;
+    }
+
+    // call CMP
+    window[cmpApiPMName](commandName, moduleCallback, cmpCallParam);
+
+    function readPostMessageResponse(event) {
+      const cmpDataPkgName = `${cmpApiName}Return`;
+      const json = (typeof event.data === 'string' && event.data.includes(cmpDataPkgName)) ? JSON.parse(event.data) : event.data;
+      if (json[cmpDataPkgName] && json[cmpDataPkgName].callId) {
+        const payload = json[cmpDataPkgName];
+
+        if (cmpCallbacks.hasOwnProperty(payload.callId)) {
+          cmpCallbacks[payload.callId](payload.returnValue);
+        }
+      }
+    }
+  }
+
+  function cmpPostResponseEventCallback(evt) {
+    if (evt) {
+      logInfo('Received a postmsg response from GPP CMP for event', evt);
```

Comment:
I was a bit unsure how to tell which parent function was calling the callback to know if I should make a cmpDirect call or an iframe call (for the getGPPData).  The callback's call signature is determined by the CMP, so there's little context to get within the callback function.  

I'm sorry if I'm missing something obvious; we can chat over slack if there's an approach you had in mind on how to unify the callback functions.

Reply:
My approach would be to implement two functions with the same signature for `callCmpWhileInFrame` and `callCmpDirect`, so that you can choose one at the beginning and everything else does not need to worry about whether frames are involved.  As I understand it the call arguments and return values do not change in the iframe case so it should be possible to make it invisible.